_drafts/conviva-dag-process-time-prod-issue-draft.md

```diff
@@ -0,0 +1,160 @@
+
+Now, let's contrast this with the approach that `ArcSwap` uses. `ArcSwap` follows the [read-copy-update (RCU)](https://docs.kernel.org/RCU/whatisRCU.html) methodology where:
+* readers access the data without locking
+* writers create a new copy of the data
+* writers atomically swap in the new data
+* old data is reclaimed later during a reclamation phase
+
+The `ArcSwap` repo even has a [method called `rcu`](https://github.com/vorner/arc-swap/blob/b12da9d783d27111d31afc77e70b07ce6acdf9f6/src/lib.rs#L603).
+
+<div class="aside">This is not unlike how [snapshot isolation](https://jepsen.io/consistency/models/snapshot-isolation) works in databases with multi-version concurrency control. The purpose is, of course, different but there are overlaps in the mechanism.<br/><br/></div>
+
+`ArcSwap` accomplishes this with a [thread-local epoch counter to track "debt"](https://github.com/vorner/arc-swap/blob/master/src/debt/list.rs#L335) which avoids cache contention issues that crop up when updating a shared read counter.
+
+A new version of the data is swapped in using the [standard `cmp_xchg`](https://github.com/vorner/arc-swap/blob/master/src/strategy/hybrid.rs#L207) operation. This marks the beginning of a new epoch, but the data associated with the old epoch isn't cleaned up until all "debt" is paid off, that is until all readers of the previous epoch have finished. Once the readers from the previous epoch have finished, the previous epoch is "reclaimed".
+
+```text
+  [Core 1]             [Core 2]             [Core 3]
+     |                   |                   |
+     | load()            | load()            | load()
+     |                   |                   |
+     v                   v                   v
+
+[Thread-Local Epoch 1] [Thread-Local Epoch 2] [Thread-Local Epoch 3]
+      (read guard)           (read guard)         (read guard)
+
+      ┌─────────────────────────────┐
+      │        ArcSwap<T>          │
+      │ ┌────────────────────────┐ │
+      │ │ Arc<T>: Current value  │ │ <── atomic ptr (no cache bouncing)
+      │ └────────────────────────┘ │
+      └─────────────────────────────┘
+
+     Writer swaps in new Arc<T> using atomic store()
+     └── Old Arc<T> placed into deferred queue
+         └── Only dropped when all read guards released
+
+✅ No shared counters for read
+✅ No cache line bouncing
+✅ Readers are wait-free and isolated
+```
+
+The big difference between a concurrent hash map and `ArcSwap` is that `ArcSwap` requires swapping out the entirety of the underlying data with every write but trades this off with very cheap reads. Writes don't even have to wait for all readers to finish since a new epoch is created with the new version of data.
+
+Hash maps on the other hand allow updating invidual portions of data in the hash map but this is where it becomes important that we have an *almost* read-only scenario because the additional overhead of writes with `ArcSwap` is worth paying here since reads are faster.
```

Comment:
In our use case the amount of data is very small so swapping out the whole thing is acceptable.

Reply:
yeah added a small note about this in 7b65223885134b32a8e1ae4e71fb18525647faa5